proof-of-spacetime.md

```diff
@@ -53,7 +53,11 @@ Each reported fault carries a penality with it.
 
 ```go
 // Generate a new PoSt.
-func GeneratePoSt(sectorSize BytesAmount, sectors SectorSet, seed Seed, faults FaultSet) PoStProof {
+// - The `sectors` slice represents the sectors in a miner's proving set and
+//   must be sorted by Commitment in lexicographical order.
+// - The `faults` slice contains indices into the `sectors` slice, each
+//   representing a previously self-reported storage fault.
+func GeneratePoSt(sectorSize BytesAmount, sectors []ReplicaInfo, seed Seed, faults []Uint) PoStProof {
```

Comment:
`GeneratePoSt` needs access to the replicas' bytes, yes?

Reply:
Yes, the bytes are needed, but… whether they need to be supplied to the function or not depends on what you `GeneratePost` represents. It's a good question, and this is the right conversation to be having.

If we think in terms of the actual SNARK proof which will be generated, the replica data is a *private* input. The proof demonstrates knowledge of that private data but without exposing the data.

Whether you think the caller of `GeneratePoSt` has access to that data depends on whether you think the caller is the 'outside world', i.e. the blockchain – or an 'inside collaborator' requesting a proof in service of its obligation to assemble a proof for the blockchain.

Obviously in order to generate a PoSt, the miner has to have access to the replica bytes. In fact, this is exactly what we are trying to prove. So whether or not those bytes has to be passed depends on whether or not the provider of `GeneratePost` is the entity expected to have the data — or whether the caller is the entity responsible for storing and providing the replica bytes (as well as merkle tree caches, etc.)

If `GeneratePoSt` is called by Sector Base or by an entity which can deal directly with Sector Base, then it should pass the replica bytes (or a handle to them) with the request. Otherwise `GeneratePoSt` is a request to prove (with zero knowledge leaked) that the callee indeed has access to those bytes, and they shouldn't be passed.

As far as I know, this distinction and the entities needed to reason about it clearly aren't really reflected in the spec yet. So whatever we decide here should eventually be supported by some treatment of Sector Base (whether or not by that name) and how it relates to both the proving subsystem and the node. If we don't want to expose the concept of this entity to the spec, then we still need to decide whether it is implicitly associated with the node or the proving subsystem. Current *implementation* aligns it with the proving subsystem (for good reason). For that reason, *absent a decision to specify it otherwise*, I think we should probably go with that assumption when specifying the proofs API.

I do know that the spec for Seal includes paths — which implies that in that case we treat the Sector *Builder* (also a concept unknown to the spec) as being on the node side of the proofs boundary.